vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md

```diff
@@ -0,0 +1,109 @@
+# AEP-4016: Support for in place updates in VPA
+
+<!-- toc -->
+- [Summary](#summary)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Context](#context)
+- [Design Details](#design-details)
+    - [Applying Updates During Pod Admission](#applying-updates-during-pod-admission)
+    - [Applying Disruption-free Updates](#Applying-disruption-free-updates)
+    - [Applying Disruptive Updates](#applying-disruptive-updates)
+    - [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+
+VPA applies its recommendations with a mutating webhook, during pod creation. It can also evict
+pods expecting that it will apply the recommendation when the pod is recreated. This is a
+disruptive process so VPA has some mechanism to avoid too frequent disruptive updates. 
+
+This proposal allows VPA to apply its recommendations more frequently, with less disruption by
+using in-place update feature which is an alpha feature available in Kubernetes 1.27.
+
+This proposal enables only core uses of  in place updates in VPA with intention of gathering more feedback. Any more
+advanced uses of in place updates in VPA (like applying different recommendations during pod initialization) will be
+introduced as separate enhancement proposals.
+
+### Goals
+
+* Allow VPA to actuate without disruption,
+* Allow VPA to actuate more frequently, when it can actuate without disruption,
+* Allow VPA to actuate in situations where actuation by eviction doesn't work.
+
+### Non-Goals
+
+* Improved handling of injected sidecars
+  * Separate AEP will improve VPAs handling of injected sidecars.
+
+## Proposal
+
+Add new supported values of
+[`UpdateMode`](https://github.com/kubernetes/autoscaler/blob/71b489f5aec3899157b37472cdf36a1de223d011/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L124):
+
+* `InPlaceOnly` and
+* `InPlaceOrRecreate`.
+
+## Context
+
+[In-place update of pod resources KEP](https://github.com/kubernetes/enhancements/issues/1287) is available in alpha in
+[Kubernetes 1.27](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#api-change-3). The
+feature allows changing container resources while the container is running. It also adds 
+[`ResizePolicy`](https://github.com/kubernetes/api/blob/8360d82aecbc72aa039281a394ebed2eaf0c0ccc/core/v1/types.go#L2448)
+field to Container. This field indicates that the container should be restarted when its resources
+
+## Design Details
+
+VPA Updater will execute in-place updates or eviction-updates in the newly added `InPlaceOnly` and `InPlaceOrRecreate`
+modes. 
+
+### Applying Updates During Pod Admission
+
+For VPAs in `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Admission Controller will apply updates to starting pods,
+like it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+
+### Applying Disruption-free Updates
```

Comment:
Are we doing something about the case where an in-place update becomes `infeasible`, because the Pod doesn't fit on this Node anymore? Otherwise it could happen that using in-place updates creates more problems for people's workloads.

We could e.g. react to events on the Pod, [which should be generated by kubelet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#kubelet-and-api-server-interaction)
> kubelet will generate Events on the Pod whenever a resize is accepted or rejected

I wasn't sure where/how to find which events should be generated in this case, though?

Reply:
FWIW I think VPA should react to both Infeasible (this is rather easy) and also to long-lasting Deferred status: "it fits on this node" might require evicting all other Pods and it might take a long time to actually achieve this (IIUC there is no controller which will just do it for VPA).

I'd like to make sure that a long-lasting Deferred status doesn't block VPA from actuating the recommendation to other Pods in the workload.

Reply:
Does the "VPA updater will consider that the update failed if" part I added answer your question? Or were you thinking about something else (scale down in place but then can't scale up in place to the old size)?

Reply:
@kgolab @voelzmo if you think I should change the doc please let me know and I'll add changes to #5877